File: 345.reverse-vowels-of-a-string.java
import java.util.ArrayList;
import java.util.Set;

/*
 * @lc app=leetcode id=345 lang=java
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start
class Solution {
    public String reverseVowels(String s) {
        int len = s.length();
        char[] arr = s.toCharArray();
        int left = 0;
        int right = len - 1;
        while (left < right) {
            char leftChar = arr[left];
            char rightChar = arr[right];

            if (isVowels(leftChar) == false) {
                left++;
            } else if (isVowels(rightChar) == false) {
                right--;
            } else {
                // leftChar and rightChar are vowel at same time
                // swap
                arr[left] = rightChar;
                arr[right] = leftChar;
                left++;
                right--;
            }
        }

        String res = String.valueOf(arr);
        return res;
    }

    private boolean isVowels(char x) {
        return (x == 'a' || x == 'e' || x == 'i' || x == 'o' || x == 'u' || x == 'A' || x == 'E' || x == 'I' || x == 'O'
                || x == 'U');
    }
}
// @lc code=end

File: 345.reverse-vowels-of-a-string.py
#
# @lc app=leetcode id=345 lang=python3
#
# [345] Reverse Vowels of a String
#

# @lc code=start

class Solution:
    def reverseVowels(self, s: str) -> str:
        # 2 pointers, left right, find vowels and swap
        chars=list(s)
        n=len(chars)
        vowels="aeiouAEIOU"
        left,right=0,n-1
        while left<right:
            if chars[left] in vowels and chars[right] in vowels:
                chars[left], chars[right]=chars[right], chars[left]
                left+=1
                right-=1
            if chars[left] not in vowels:
                left+=1
            if chars[right] not in vowels:
                right-=1
        
        return "".join(chars)
        
# @lc code=end

